src/coreclr/scripts/cse_ml/evaluate.py

```diff
@@ -0,0 +1,225 @@
+        original = superpmi.jit_method(m_id, JitMetrics=1)
+        no_cse = superpmi.jit_method(m_id, JitMetrics=1, JitRLHook=1, JitRLHookCSEDecisions=[])
+
+        if original is None or no_cse is None:
+            set_result(data, m_id, 0, 0, 0, ModelResult.JIT_FAILED)
+            continue
+
+        choices = []
+        results = []
+        while True:
+            prev_method = results[-1] if results else no_cse
+
+            # If we have no more CSEs to apply, we are done.  We expect this not to happen on the first
+            # iteration because we filter out methods that have no CSEs to apply.
+            if not any(x.can_apply for x in prev_method.cse_candidates):
+                set_result(data, m_id, original.perf_score, no_cse.perf_score, prev_method.perf_score)
+
+                assert choices  # We must have made at least one selection
+                break
+
+            action = get_most_likley_allowed_action(jitrl, prev_method, choices)
+            if action is None:
+                set_result(data, m_id, original.perf_score, no_cse.perf_score, prev_method.perf_score)
+                break
+
+            # apply the CSE
+            choices.append(action)
+            new_method = superpmi.jit_method(m_id, JitMetrics=1, JitRLHook=1, JitRLHookCSEDecisions=choices)
+            if new_method is None:
+                set_result(data, m_id, original.perf_score, no_cse.perf_score, prev_method.perf_score,
+                            ModelResult.JIT_FAILED)
+                break
+
+            results.append(new_method)
+
+            # mark choices as applied
+            for c in choices:
+                new_method.cse_candidates[c].applied = True
+
+    return pandas.DataFrame(data)
+
+def evaluate(superpmi, jitrl, methods, model_name, csv_file) -> pandas.DataFrame:
+    """Evaluate the model and save to the specified CSV file."""
+    if os.path.exists(csv_file):
+        return pandas.read_csv(csv_file)
+
+    result = test_model(superpmi, jitrl, methods, model_name)
+    result.to_csv(csv_file)
+    return result
+
+def enumerate_models(dir_or_file):
+    """Enumerates the models in the specified directory."""
+    if os.path.isfile(dir_or_file):
+        return [dir_or_file]
+
+    def extract_number(file):
+        return int(file.split("_")[-1]) if file.split("_")[-1].isdigit() else 100000000
+
+    files = [os.path.splitext(file)[0] for file in os.listdir(dir_or_file) if file.endswith(".zip")]
+    return sorted(files, key=extract_number, reverse=True)
+
+def print_result(result, model, kind):
+    """Prints the results."""
+
+    print('-' * 40 + f" {model} results " + '-' * 40)
+    print()
+
+    print(f"{kind} results:")
+    print()
+
+    print("Comparisons:")
+    no_jit_failure = result[result['failed'] != ModelResult.JIT_FAILED]
+
+    # next calculate how often we improved on the heuristic
+    improved = no_jit_failure[no_jit_failure['model_score'] < no_jit_failure['heuristic_score']]
```

Comment:
I think this touches on how the aggregate score is computed.

Generally I like to use the geomean. If we have $N$ methods and for each have base score $b_i$ and diff score $d_i$. Then the aggregate geomean $G$ is

$$ G = e^{{1\over{N}} \sum_i log(d_i/b_i)} $$

(here lower is better) and I expect the "best possible" improvement to be around 0.99 (my policy gets about 0.994).

Reply:
Ah interesting.  I will add this to the next update, thanks for the suggestion!